packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/props/cache.rs

```diff
@@ -0,0 +1,254 @@
+/// that intelligently recomputes props when their dependencies change.
+#[derive(Debug)]
+struct CachedProp {
+    /// Cached value of the prop. Will be `None` if the prop hasn't been computed or is stale.
+    value: Rc<RefCell<Option<PropValue>>>,
+    meta: RefCell<CachedPropMeta>,
+}
+
+impl CachedProp {
+    pub fn new() -> Self {
+        CachedProp {
+            value: Rc::new(RefCell::new(None)),
+            meta: RefCell::new(CachedPropMeta {
+                status: PropStatus::Unresolved,
+                came_from_default: false,
+                change_counter: 0,
+            }),
+        }
+    }
+
+    /// Sets the value and marks it as fresh using internal mutability.
+    /// **For internal use only.**
+    fn _set_value(&self, value: PropValue) {
+        *self.value.borrow_mut() = Some(value);
+        self.meta.borrow_mut().status = PropStatus::Fresh;
+        let change_counter = self.meta.borrow().change_counter;
+        // A little bit of safety in case someone wiggles their mouse 4 billion times
+        self.meta.borrow_mut().change_counter = change_counter.wrapping_add(1);
+    }
+
+    /// Sets the prop as stale and clears the cached value using interior mutability.
+    /// **For internal use only.**
+    fn _set_stale(&self) {
+        *self.value.borrow_mut() = None;
+        self.meta.borrow_mut().status = PropStatus::Stale;
+    }
+
+    /// Set `came_from_default` using interior mutability.
+    /// **For internal use only.**
+    fn _set_came_from_default(&self, came_from_default: bool) {
+        self.meta.borrow_mut().came_from_default = came_from_default;
+    }
+
+    /// Set `status` using interior mutability.
+    /// **For internal use only.**
+    fn set_status(&self, status: PropStatus) {
+        self.meta.borrow_mut().status = status;
+    }
+
+    /// Get the value of the prop. If the cached value is not fresh,
+    /// the `calculate` function will be called to get the value.
+    ///
+    /// This function uses interior mutability to cache computed values.
+    pub fn get_value<CalculateFn: Fn() -> PropCalcResult<PropValue>>(
+        &self,
+        calculate: CalculateFn,
+    ) -> Ref<PropValue> {
+        match self.get_status() {
+            PropStatus::Fresh => self.get_cached_value().expect(
+                "Prop is marked as Fresh but no value is cached. This state should be unreachable",
+            ),
+            PropStatus::Stale | PropStatus::Resolved => {
+                let result = calculate();
+                match result {
+                    PropCalcResult::Calculated(value) => {
+                        self._set_value(value);
+                    }
+                    PropCalcResult::FromDefault(value) => {
+                        self._set_value(value);
+                        self._set_came_from_default(true);
+                    }
+                    PropCalcResult::NoChange => {
+                        self.set_status(PropStatus::Fresh);
```

Comment:
For this `set_status` to work, you'll need to change `_set_stale` to just change the status and not clear the value.

Also, if the `status` is `Resolved`, it should panic, as it means the prop author made a mistake.